README.md
# NAME

Pod::Elemental::Transformer::Splint - Documentation from class metadata

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.10.1+-blue.svg" alt="Requires Perl 5.10.1+" />
    <a href="http://cpants.cpanauthors.org/release/CSSON/Pod-Elemental-Transformer-Splint-0.1202"><img src="http://badgedepot.code301.com/badge/kwalitee/CSSON/Pod-Elemental-Transformer-Splint/0.1202" alt="Distribution kwalitee" /></a>
    <a href="http://matrix.cpantesters.org/?dist=Pod-Elemental-Transformer-Splint%200.1202"><img src="http://badgedepot.code301.com/badge/cpantesters/Pod-Elemental-Transformer-Splint/0.1202" alt="CPAN Testers result" /></a>
    <img src="https://img.shields.io/badge/coverage-81.4%-orange.svg" alt="coverage 81.4%" />
    </p>
</div>

# VERSION

Version 0.1202, released 2020-12-26.

# SYNOPSIS

    # in weaver.ini
    [-Transformer / Splint]
    transformer = Splint

# DESCRIPTION

Pod::Elemental::Transformer::Splint uses [MooseX::AttributeDocumented](https://metacpan.org/pod/MooseX::AttributeDocumented) to add inlined documentation about attributes to pod.
If you write your classes with [Moops](https://metacpan.org/pod/Moops) you can also document method signatures with [Kavorka::TraitFor::Parameter::doc](https://metacpan.org/pod/Kavorka::TraitFor::Parameter::doc) (and [::ReturnType::doc](https://metacpan.org/pod/Kavorka::TraitFor::ReturnType::doc)).

A class defined like this:

    package My::Class;

    use Moose;

    has has_brakes => (
        is => 'ro',
        isa => Bool,
        default => 1,
        traits => ['Documented'],
        documentation => 'Does the bike have brakes?',
        documentation_alts => {
            0 => 'Hopefully a track bike',
            1 => 'Always nice to have',
        },
    );

    =pod

    :splint classname My::Class

    :splint attributes

    =cut

Will render like this (to html):

_begin_

_end_

A [Moops](https://metacpan.org/pod/Moops) class defined like this:

    class My::MoopsClass using Moose {

        ...

        method advanced_method(Int $integer                        does doc("Just an integer\nmethod_doc|This method is advanced."),
                               ArrayRef[Str|Bool] $lots_of_stuff   does doc('It works with both types'),
                               Str :$name!                         does doc("What's the name"),
                               Int :$age                           does doc('The age of the thing') = 0,
                               Str :$pseudonym                     does doc('Incognito..')
                           --> Bool but assumed                    does doc('Did it succeed?')

        ) {
            return 1;
        }

        method less_advanced($stuff,
                             $another_thing                     does doc("Don't know what we get here"),
                             ArrayRef $the_remaining is slurpy  does doc('All the remaining')
        )  {
            return 1;
        }

        ...
    }

    =pod

    :splint classname My::MoopsClass

    :splint method advanced_method

    It needs lots of documentation.

    :splint method less_advanced

    =cut

Will render like this (to html):

_begin_

_end_

# ATTRIBUTES

The following settings are available in `weaver.ini`:

## command\_name

Default: `:splint`

Defines the command used at the beginning of the line in pod.

## attribute\_renderer

Default: `HTML=HtmlDefault, markdown=HtmlDefault`

Define which renderers to use. Comma separated list of pairs where the key defines the format in pod and the value defines the renderer (in the `Pod::Elemental::Transformer::Splint::AttributeRenderer` namespace).

The default will render each attribute like this:

    =begin HTML

    <!-- attribute information -->

    =end HTML

    =begin markdown

    <!-- attribute information -->

    =end markdown

## method\_renderer

Default: `HTML=HtmlDefault, markdown=HtmlDefault`

Similar to ["attribute\_renderer"](#attribute_renderer) but for methods. This is currently only assumed to work for methods defined with [Kavorka](https://metacpan.org/pod/Kavorka) or [Moops](https://metacpan.org/pod/Moops).

Method renderers are in the `Pod::Elemental::Transformer::Splint::MethodRenderer` namespace.

## type\_libraries

Default: `undef`

If you use [Type::Tiny](https://metacpan.org/pod/Type::Tiny) based type libraries, the types are usually linked to the correct library. Under some circumstances it might be necessary to specify which library a certain type
belongs to.

It is a space separated list:

    type_libraries = Custom::Types=AType Types::Standard=Str,Int

## default\_type\_library

Default: `Types::Standard`

The default Type::Tiny based type library to link types to.

# SOURCE

[https://github.com/Csson/p5-Pod-Elemental-Transformer-Splint](https://github.com/Csson/p5-Pod-Elemental-Transformer-Splint)

# HOMEPAGE

[https://metacpan.org/release/Pod-Elemental-Transformer-Splint](https://metacpan.org/release/Pod-Elemental-Transformer-Splint)

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Pod-Elemental-Transformer-Splint

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.1202 2020-12-26T17:05:06Z UTC
 [Other Changes]
 - Explicitly set $VERSION
 - Replace TravisCI with Github Actions

 [Requirements]
 - Added missing req Pod::Elemental::Transformer

 [Stats & Dependency Changes]
 - 8 files changed, 62 insertions(+), 6 deletions(-)
 - (run req) + Pod::Elemental::Transformer 0.103004
 - (run req) ~ List::MoreUtils (any) --> 0.408
 - (run req) ~ List::Util (any) --> 1.45
 - (run req) ~ Module::Load (any) --> 0.26
 - (run req) ~ Path::Tiny (any) --> 0.072
 - (run req) ~ Safe::Isa (any) --> 1.000005
 - (run req) ~ Try::Tiny (any) --> 0.24
 - (run req) ~ Types::Standard (any) --> 1.000006
 - (run req) ~ namespace::autoclean (any) --> 0.22
 - (dev req) + Badge::Depot 0.0103
 - (dev req) + Badge::Depot::Plugin::Coverage (any)
 - (dev req) + Badge::Depot::Plugin::Cpantesters (any)
 - (dev req) + Badge::Depot::Plugin::Kwalitee (any)
 - (dev req) + Badge::Depot::Plugin::Perl (any)
 - (dev req) + Dist::Zilla 6.015
 - (dev req) +
   Dist::Zilla::Plugin::Author::CSSON::GithubActions::Workflow::TestWithMakefile
   0.0106
 - (dev req) + Dist::Zilla::Plugin::Authority 1.009
 - (dev req) + Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   0.009
 - (dev req) + Dist::Zilla::Plugin::ChangeStats::Dependencies::Git 0.0200
 - (dev req) + Dist::Zilla::Plugin::Git 2.046
 - (dev req) + Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.014
 - (dev req) + Dist::Zilla::Plugin::Git::Commit (any)
 - (dev req) + Dist::Zilla::Plugin::MetaConfig (any)
 - (dev req) + Dist::Zilla::Plugin::NextRelease::Grouped 0.0200
 - (dev req) + Dist::Zilla::Plugin::PromptIfStale 0.057
 - (dev req) + Dist::Zilla::Plugin::RewriteVersion::Transitional 0.007
 - (dev req) + Pod::Weaver::Section::GenerateSection 1.01
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::Version 1
 - (dev req) ~ Dist::Zilla::Plugin::CheckChangesHasContent (any) --> 0.011
 - (dev req) ~ Dist::Zilla::Plugin::Clean (any) --> 0.07
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild (any) --> 0.170880
 - (dev req) ~ Dist::Zilla::Plugin::Git::Contributors (any) --> 0.035
 - (dev req) ~ Dist::Zilla::Plugin::GithubMeta (any) --> 0.54
 - (dev req) ~ Dist::Zilla::Plugin::InstallRelease (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::MetaProvides::Package (any) -->
   2.004003
 - (dev req) ~ Dist::Zilla::Plugin::PodWeaver (any) --> 4.006
 - (dev req) ~ Dist::Zilla::Plugin::ReadmeAnyFromPod (any) --> 0.163250
 - (dev req) ~ Dist::Zilla::Plugin::RunExtraTests (any) --> 0.029
 - (dev req) ~ Dist::Zilla::Plugin::Test::Compile (any) --> 2.058
 - (dev req) ~ Dist::Zilla::Plugin::Test::EOF (any) --> 0.0501
 - (dev req) ~ Dist::Zilla::Plugin::Test::EOL (any) --> 0.18
 - (dev req) ~ Dist::Zilla::Plugin::Test::NoTabs (any) --> 0.15
 - (dev req) ~ Dist::Zilla::Plugin::Test::ReportPrereqs (any) --> 0.027
 - (dev req) ~ Dist::Zilla::Plugin::Test::Version (any) --> 1.09
 - (dev req) ~ Pod::Weaver::Section::Badges (any) --> 0.0401
 - (dev req) ~ Test::NoTabs (any) --> 1.4
 - (dev req) ~ Test::Warnings (any) --> 0.026
 - (dev req) - Dist::Iller::Config::Author::CSSON
 - (dev req) - Dist::Zilla::Plugin::MetaProvides::Class
 - (dev req) - Dist::Zilla::Plugin::NextRelease
 - (dev req) - Dist::Zilla::Plugin::NextVersion::Semantic
 - (dev req) - Dist::Zilla::Plugin::OurPkgVersion
 - (dev req) - Dist::Zilla::Plugin::PodnameFromClassname
 - (dev req) - Dist::Zilla::Plugin::Prereqs::Plugins
 - (dev req) - Dist::Zilla::Plugin::PreviousVersion::Changelog
 - (dev req) - Dist::Zilla::Plugin::ReversionOnRelease
 - (dev req) - Dist::Zilla::Plugin::Test::Kwalitee::Extra
 - (dev req) - Dist::Zilla::Plugin::TravisYML
 - (dev req) - Pod::Weaver::Section::Homepage::DefaultCPAN
 - (dev req) - Pod::Weaver::Section::Source::DefaultGitHub
 - (dev sug) + Dist::Iller 0.1409
 - (dev sug) + Dist::Iller::Config::Author::CSSON 0.0327

0.1201 2016-02-03T10:27:59Z UTC
 [Documentation]
 - Remove empty section.

 [Requirements]
 - Require List::Util+List::MoreUtils instead of List::AllUtils.

0.1200 2016-02-02T21:11:41Z UTC
 [Requirements]
 - Now works 5.10+ (skips tests that require 5.14).
 - Kavorka::TraitFor::Parameter::doc downgraded from recommends to
   suggests.

0.1100 2016-01-12T22:17:40Z UTC
 [API Changes]
 - Breaking change: html_attribute_renderer is replaced by
   attribute_renderer.
 - Breaking change: html_method_renderer is replaced by method_renderer.

 [Documentation]
 - Document weaver.ini configuration options.

 [New Features]
 - The new render settings in weaver.ini makes it possible to render to
   many formats.

 [Other Changes]
 - ::HtmlRenderer: Minor improvements to rendered html.

 [Requirements]
 - New prereq: Module::Load
 - New prereq: Types::Standard
 - New prereq: Ref::Util

0.1003 2016-01-11T21:39:31Z UTC
 [Other Changes]
 - Converted to use latest Dist::Iller::Config::Author::CSSON

0.1002 2015-01-18T00:37:18Z CET
 [Enhancements]
 - Improved the order that attributes are rendered.
 - Improve output for method parameters that are hash or array refs, and
   have a default value.
 - Ensure whitespace before [ in type definitions for better legibility.
 - AttributeRenderer/HtmlDefault: Add a note if an attribute is not
   available in the constructor.
 - MethodRenderer/HtmlDefault: Use (mildly) different colours in the three
   headings

 [New Features]
 - (Undocumented feature: ':splint methods' will try to render metadata on
   all methods.)

 [Requirements]
 - Added new requirements List::UtilsBy and Try::Tiny.
 - Bump required version of Syntax-Feature-Qs to 0.2003.

0.1001 2015-01-14T03:06:07Z CET
 [Documentation]
 - Fixed Moops class example.

 [Enhancements]
 - Uses 'pseudo colspan'. Metacpan seem to remove colspan attributes.

 [Meta]
 - Fixed test prereqs.
 - Fixed no-op test to actually test.

Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.017.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Documentation from class metadata",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Pod-Elemental-Transformer-Splint",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "Pod::Elemental::Transformer::Splint",
  "PREREQ_PM" => {
    "List::MoreUtils" => "0.408",
    "List::Util" => "1.45",
    "List::UtilsBy" => 0,
    "Module::Load" => "0.26",
    "Moose" => "2.00",
    "Moose::Role" => "2.1400",
    "Path::Tiny" => "0.072",
    "Pod::Elemental::Transformer" => "0.103004",
    "Ref::Util" => "0.007",
    "Safe::Isa" => "1.000005",
    "Syntax::Feature::Qs" => "0.2003",
    "Try::Tiny" => "0.24",
    "Types::Standard" => "1.000006",
    "namespace::autoclean" => "0.22",
    "syntax" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Pod::Elemental" => 0,
    "Pod::Elemental::Transformer::Pod5" => 0,
    "Test::Differences" => 0,
    "Test::More" => "0.96",
    "Test::Warnings" => 0
  },
  "VERSION" => "0.1203",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "List::MoreUtils" => "0.408",
  "List::Util" => "1.45",
  "List::UtilsBy" => 0,
  "Module::Load" => "0.26",
  "Moose" => "2.00",
  "Moose::Role" => "2.1400",
  "Path::Tiny" => "0.072",
  "Pod::Elemental" => 0,
  "Pod::Elemental::Transformer" => "0.103004",
  "Pod::Elemental::Transformer::Pod5" => 0,
  "Ref::Util" => "0.007",
  "Safe::Isa" => "1.000005",
  "Syntax::Feature::Qs" => "0.2003",
  "Test::Differences" => 0,
  "Test::More" => "0.96",
  "Test::Warnings" => 0,
  "Try::Tiny" => "0.24",
  "Types::Standard" => "1.000006",
  "namespace::autoclean" => "0.22",
  "syntax" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
